### DISCUSSION TITLE
Module debugging

### DISCUSSION THREAD

**Discussion by markagama** [oqtane/oqtane.framework#3144] [General] [answered]
I'm following @ADefWebserver example for Syncfusion and thx to @sbwalker with the fix to the IHostResources issue, I can now see the grid and get an error about an unknown 'source' parameter.

When debugging a module, should I debug using the module opened in VS2022 or the oqtane opened in VS2022? This is a bit perplexing, but hopefully a simple query.

When I get @ADefWebserver example working fully, I'll make the source code available so that others can save time when working with Syncfusion controls and Oqtane 4.

**Comment by sbwalker** [comment]
@markagama I am not sure what you used as a starting point... but if you used the Create Module option from within the framework to scaffold the code for your module (which is the recommended approach) then you will have a module solution which includes the Oqtane.Server project as a reference. You need to ensure the Oqtane.Server project is set as the Startup Project in Visual Studio. Then when you run the module solution from Visual Studio, the Oqtane Framework will load automatically and you will also be able to hit breakpoints in your module code. Using this approach you do not need to open the Oqtane solution at all. The only other thing you will need to remember is that when you make changes to your module code you will need to execute a Build in Visual Studio to ensure the  the module DLLs get deployed to the appropriate locations. After you Build you can Run and your changes will be reflected in the application.

**Reply by markagama** [reply]
thx @sbwalker for replying. I created a mdule in Oqtane 4.2 and tried to modify the code to use Michael's Oqtane 2 HelpDesk example. It was the only syncfusion example that I could find. As mentioned below, with the resource fix, I now have the control on the page, and it appears that the problem is the empty table being returned - the db calls are likely to be using code that has been updated for Oqtane 4. I should have some time later this week to continue exploring. As there are no other examples, when I get this working, I'll make the code available for others, as an example is always welcome.

**Reply by sbwalker** [reply]
In your Client project in the Service.cs class there is a method for Get{YourModuleName}s. The latest template uses the following code to avoid an error on an empty table ( the main difference is the inclusion of the Enumerable.Empty<Models.{YourModuleName}>().ToList() parameter):

```
        public async Task<List<Models.{YourModuleName}>> Get{YourModuleName}sAsync(int ModuleId)
        {
            List<Models.{YourModuleName}> {YourModuleName}s = await GetJsonAsync<List<Models.{YourModuleName}>>(CreateAuthorizationPolicyUrl($"{Apiurl}?moduleid={YourModuleName}", EntityNames.Module, ModuleId), Enumerable.Empty<Models.{YourModuleName}>().ToList());
            return {YourModuleName}s.OrderBy(item => item.Name).ToList();
        }
```

**Reply by markagama** [reply]
thx @sbwalker I can see what this does, and with this change Michael's module still has an error that must occur further upstream in the code. I'll start again with a simple module that displays the control, then build a database layer to go with this based on the template example.

**Comment by sbwalker** [comment, marked as answer]
An error for "unknown 'source' parameter" usually means that the database table related to the API is empty. In 4.0.2 the module template was modified to use a new method where an empty list can be returned as a default in these scenarios - to prevent an error.

**Reply by markagama** [reply]
thx @sbwalker I've got the control on the page now that you alerted me to the resource fix and have started to look at the data side of Michaels' example. I'm looking at the code that was created in the default module and the differences with Michael's Oqtane 2 HelpDesk example.
